### PR TITLE
Parity with Java for Metrics

### PR DESCRIPTION
**Pull request by cwe1ss**
I'm comparing the C# code with the Java code and I'm trying to get it as close as possible to make future changes easier. (#29)

This PR brings the "Metrics" feature closer to Java. However there still need to be a few differences:
* Java has a concrete class called `Metrics` which is used everywhere. Having a class with this name is not nice in C# because the namespace has the same name and people would have to use `Metrics.Metrics`. As I couldn't come up with a different name, I just kept the already existing `IMetrics` interface and moved Java's logic to a new `MetricsImpl` class.
* C# can't use complex types on attribute constructors. The current C# code already used some custom logic for this but I reverted this and simplified it. It's now pretty close to Java.

I also replaced the metrics unit tests with the current version from Java.

### CODE REVIEW

File: src/Jaeger.Core/Metrics/MetricsImpl.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+
+namespace Jaeger.Core.Metrics
+{
+    public class MetricsImpl : IMetrics
+    {
+        /// <inheritdoc/>
+        [Metric(name: "traces", tag1Key: "state", tag1Value: "started", tag2Key: "sampled", tag2Value: "y")]
```
Comment (yurishkuro) [inline src/Jaeger.Core/Metrics/MetricsImpl.cs]: Is this a limitation of the annotation syntax that you can't do a dictionary?

If so I would recommend just having a string like `tags="state=started, sampled=y"` and parse it during construction. The objective is to make the annotation readable & easy, not to help the parsing engine (which isn't that complicated anyway).

File: src/Jaeger.Core/Reporters/RemoteReporter.cs [hunk 2]
```diff
@@ -28,7 +28,11 @@ private RemoteReporter(ITransport transport, ILoggerFactory loggerFactory, IMetr
             try
             {
                 int n = await _transport.CloseAsync(CancellationToken.None).ConfigureAwait(false);
-                _metrics.ReporterSuccess.Inc(n);
+
+                if (n > 0)
```
Comment (yurishkuro) [inline src/Jaeger.Core/Reporters/RemoteReporter.cs]: does it matter if it's 0? seems like unnecessary complication of the code

File: test/Jaeger.Core.Tests/Reporters/RemoteReporterTests.cs [hunk 3]
```diff
@@ -67,27 +67,15 @@ public void RemoteReporter_ShouldCallMetrics()
...
-        public void RemoteReporter_ShouldCallMetricsFactory()
-        {
-            var transport = Substitute.For<ITransport>();
-            var metricsFactory = InMemoryMetricsFactory.Instance;
-
-            using (var reporter = new RemoteReporter.Builder(transport).WithMetricsFactory(metricsFactory).Build())
-            {
-                Assert.IsType<InMemoryMetricsFactory.InMemoryElement>(reporter._metrics.ReporterSuccess);
-            }
+            metrics.ReporterSuccess.Received(1).Inc(2);
```
Comment (yurishkuro) [inline test/Jaeger.Core.Tests/Reporters/RemoteReporterTests.cs]: why not use in-memory implementation and just assert the final counter value? With a mock you are testing not only external side effect, i.e. the resulting value of the counter which is what we're interested in, but also _how it gets there_, which is mostly an irrelevant implementation detail that makes the test itself more brittle.

File: src/Jaeger.Core/Metrics/MetricsImpl.cs (hunk 1 above)
Comment (cwe1ss) [reply src/Jaeger.Core/Metrics/MetricsImpl.cs]: Yes, this is a limitation in C#! I'll change it to your recommendation.

**Comment by cwe1ss** [comment]
Superseded by #55.

**Pull request closed without merging**